schema/elements.go

```diff
@@ -250,14 +257,49 @@ func (s *Schema) FindNamedType(name string) (TypeDef, bool) {
 // This allows callers to not care about the difference between a (possibly
 // inlined) reference and a definition.
 func (s *Schema) Resolve(tr TypeRef) (Atom, bool) {
+	result := Atom{}
+
 	if tr.NamedType != nil {
 		t, ok := s.FindNamedType(*tr.NamedType)
 		if !ok {
 			return Atom{}, false
 		}
-		return t.Atom, true
+
+		result = t.Atom
+	} else {
+		result = tr.Inlined
 	}
-	return tr.Inlined, true
+
+	// Allow field-level electives to override the referred type's modifiers
+	switch {
+	case result.Map != nil:
+		mapCopy := *result.Map
+
+		if tr.ElementRelationship != nil {
+			mapCopy.ElementRelationship = *tr.ElementRelationship
+		}
+
+		result.Map = &mapCopy
+	case result.List != nil:
+		listCopy := *result.List
+
+		if tr.ElementRelationship != nil {
+			listCopy.ElementRelationship = *tr.ElementRelationship
+		}
+
+		result.List = &listCopy
+
+	case result.Scalar != nil:
+		if tr.ElementRelationship != nil {
+			return Atom{}, false
+		}
+
+	default:
+		// Atom violates invariant that one of Map/List/Scalar is set
```

Comment:
This empty Atom return value is just terrible. I don't think we handle it properly anywhere and usually leads to hard to debug errors. Anyways, just ranting :-)

Reply:
I think the contract of this function is to not inspect the first return value of the second is `false`.